### PR TITLE
[JAX] Fix grouped GEMM error on CUDA 12.9.1 & later

### PR DESCRIPTION
**Pull request by huanghua1994**
# Description

On CUDA 12.9.1 and later versions, cuBLAS has some breaking changes and makes JAX grouped GEMM fail. This PR fixes the issues.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

1. For FP8 grouped GEMM on Hopper, ensures that the pointer to each scaling factor is 256B aligned (cuBLAS requires at least 16B alignment, use 256B to simplify the code)
2. For MXFP8 on Blackwell, adjust the `value_and_grad` output in unit test to avoid  NaN values due to overflow.

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/jax/csrc/extensions/gemm.cpp [hunk 1]
```diff
@@ -86,6 +107,23 @@ Error_Type GroupedGemmFFI(cudaStream_t stream, Buffer_Type lhs_data, Buffer_Type
   size_t bias_dtype_bytes = te_dtype_bytes(bias_dtype);
   size_t out_dtype_bytes = te_dtype_bytes(out_dtype);
 
+  if (is_tensor_scaling) {
+    cudaStream_t stream_0 = nvte_get_compute_stream(0);
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/csrc/extensions/gemm.cpp:111]: Does JAX allow using a stream separate from the one provided as a param in the FFI interface? Maybe the grouped gemm is already using these streams internally in TE common so it's okay

File: transformer_engine/jax/cpp_extensions/gemm.py [hunk 2]
```diff
@@ -98,15 +98,25 @@ def abstract(
...
         workspace_size += 256
-        workspace_size += lhs_scale_inv_aval.size + 256
-        workspace_size += rhs_scale_inv_aval.size + 256
+        if (
+            scaling_mode == ScalingMode.DELAYED_TENSOR_SCALING
+            or scaling_mode == ScalingMode.CURRENT_TENSOR_SCALING
+        ):
+            # For tensor scaling, each matrix has a single scale value, but it
+            # needs to be aligned to 16 bytes for CUDA 12.9.1 and later. We use
+            # a 256 bytes alignment to simplify the code.
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/cpp_extensions/gemm.py]: What is the additional complexity that would be needed to handle 16-byte alignment of scales instead of 256-byte alignment? I'd think it'd be better to align smaller, though at the size of one scale per typically pretty large tensor the difference between 16-byte/256-bytes per scale might be negligible.

File: transformer_engine/jax/csrc/extensions/gemm.cpp [hunk 3]
```diff
@@ -86,6 +107,23 @@ Error_Type GroupedGemmFFI(cudaStream_t stream, Buffer_Type lhs_data, Buffer_Type
...
+    size_t dpitch = tensor_scaling_sinv_aligment;
+    size_t spitch = lhs_sinv_dtype_bytes;
+    size_t width = lhs_sinv_dtype_bytes;
+    size_t height = lhs_sinv_size;
+    cudaMemcpy2DAsync(lhs_scatter_aligned_ptr, dpitch, lhs_sinv_ptr, spitch, width, height,
+                      cudaMemcpyHostToDevice, stream_0);
```
Comment (phu0ngng) [inline transformer_engine/jax/csrc/extensions/gemm.cpp]: I'm surprised that `lhs_sinv_ptr` could be a host pointer as we allocate it as `jax.ndarray`, which should be a device pointer.

File: transformer_engine/jax/csrc/extensions/gemm.cpp (hunk 1 above)
Comment (phu0ngng) [reply transformer_engine/jax/csrc/extensions/gemm.cpp:111]: No problem if we do it in this way.
However, I suspect that the `nvte_get_compute_stream(0)` is the same stream as the one given by the FFI. In this case, we should simply use the `stream` given by FFI for the code simplification.

File: transformer_engine/jax/cpp_extensions/gemm.py (hunk 2 above)
Comment (huanghua1994) [reply transformer_engine/jax/cpp_extensions/gemm.py]: Fixed, now using 16B alignment.

File: transformer_engine/jax/csrc/extensions/gemm.cpp (hunk 3 above)
Comment (huanghua1994) [reply transformer_engine/jax/csrc/extensions/gemm.cpp]: I think all `Buffer_Type` arrays from FFI are device ptr, just like the issue we have on `group_sizes`.

Comment (phu0ngng) [reply transformer_engine/jax/csrc/extensions/gemm.cpp]: If `lhs_sinv_ptr` is already a device pointer, why do we need to do a `HostToDevice` here?

Comment (huanghua1994) [reply transformer_engine/jax/csrc/extensions/gemm.cpp]: Thanks for reminding me again. I misread the previous comment. Fixed now.

**Review by phu0ngng — DISMISSED** [review]
LGTM

**Comment by phu0ngng** [comment]
Pipeline #31288765 with the nightly internal image.

All the `grouped_gemm` related tests passed. The failures on L40 and the test_multiprocessing_encoder on B200 are not related to grouped_gemm.

Thus, the PR is good to merge.

File: transformer_engine/jax/cpp_extensions/gemm.py [hunk 4]
```diff
@@ -98,15 +98,27 @@ def abstract(
...
         # necessarily 256 bytes aligned, we add some padding to ensure alignment.
-        # We also pad scale_inv swizzle buffers size for 256 bytes alignment.
-        workspace_size += 256
-        workspace_size += lhs_scale_inv_aval.size + 256
-        workspace_size += rhs_scale_inv_aval.size + 256
+        workspace_size += workspace_alignment_padding
+        if scaling_mode in (
+            ScalingMode.DELAYED_TENSOR_SCALING.value,
+            ScalingMode.CURRENT_TENSOR_SCALING.value,
+        ):
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/cpp_extensions/gemm.py:113]: nit SR: `ScalingMode(scaling_mode).is_tensor_scaling()`

File: transformer_engine/jax/quantize/tensor.py [hunk 5]
```diff
@@ -142,9 +142,9 @@ def __post_init__(self):
             pad_width = tuple(
                 (0, a - b) for a, b in zip(expected_scale_shape, expected_unpadded_scale_shape)
             )
-            # This actually pad scale_inv with nan, should we pad it with 127 directly instead?
+            # padding with the smallest number it can present
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/quantize/tensor.py:145]: Just for curiosity, was this needed to fix an issue? Or just using this PR as a chance to update it since we're touching gemm padding already

**Review by jberchtold-nvidia — APPROVED** [review]

File: transformer_engine/jax/quantize/tensor.py (hunk 5 above)
Comment (phu0ngng) [reply transformer_engine/jax/quantize/tensor.py:145]: We need this fix so that the MXFP8 GEMM with custom call works correctly.
We did not see this issue before, as without `/ jnp.sqrt(x.size)` the gradient is probably too large that it gets clipped, so all the tests passed in the past. https://github.com/NVIDIA/TransformerEngine/blob/279d0be059c774c146532cc0a09baddab4ccea5e/tests/jax/test_custom_call_compute.py#L1345-L1349
